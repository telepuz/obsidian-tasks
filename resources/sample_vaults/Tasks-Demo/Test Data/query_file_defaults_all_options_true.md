---
TQ_explain: true
TQ_extra_instructions: |-
  # press shift-return to add new lines
  not done
TQ_short_mode: true
TQ_show_backlink: true
TQ_show_cancelled_date: true
TQ_show_created_date: true
TQ_show_depends_on: true
TQ_show_done_date: true
TQ_show_due_date: true
TQ_show_edit_button: true
TQ_show_id: true
TQ_show_on_completion: true
TQ_show_postpone_button: true
TQ_show_priority: true
TQ_show_recurrence_rule: true
TQ_show_scheduled_date: true
TQ_show_start_date: true
TQ_show_tags: true
TQ_show_task_count: true
TQ_show_tree: true
TQ_show_urgency: true
---

# query_file_defaults_all_options_true

- [ ] #task Task in 'query_file_defaults_all_options_true'
