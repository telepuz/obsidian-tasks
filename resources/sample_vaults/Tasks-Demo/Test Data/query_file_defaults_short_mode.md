---
TQ_short_mode: true
---

# query_file_defaults_short_mode

- [ ] #task Task in 'query_file_defaults_short_mode'
