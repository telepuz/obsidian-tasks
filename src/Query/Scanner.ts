import { Statement } from './Statement';

function endsWith1Slash(inputLine: string) {
    return inputLine.endsWith('\\');
}

function endsWith2Slashes(inputLine: string) {
    return inputLine.endsWith('\\\\');
}

function stripLeadingWhitespace(adjustedInputLine: string) {
    return adjustedInputLine.replace(/^[ \t]*/, '');
}

function stripEndingSlashAndPrecedingWhitespace(adjustedInputLine: string) {
    return adjustedInputLine.replace(/[ \t]*\\$/, '');
}

function adjustLine(inputLine: string, continuePreviousLine: boolean) {
    let adjustedLine = inputLine;
    if (continuePreviousLine) {
        // The new line will be appended to the previous one,
        // so discard any leading white space:
        adjustedLine = stripLeadingWhitespace(inputLine);
    }

    if (endsWith2Slashes(adjustedLine)) {
        // This has at least 2 backslashes at the end of the line,
        // so convert '\\' to '\':
        adjustedLine = adjustedLine.slice(0, -1);
    } else if (endsWith1Slash(inputLine)) {
        // This is a continuation line, so remove its trailing backslash
        // and any spare white space beforehand.
        adjustedLine = stripEndingSlashAndPrecedingWhitespace(adjustedLine);
    }
    return adjustedLine;
}

/**
 * Removes newlines escaped by a backslash.
 * A trailing backslash at the end of a line can be escaped by doubling it.
 *
 * @param input input string
 * @returns modified input, as a string
 *
 * @see continueLines
 */
export function continueLinesFlattened(input: string): string {
    return continueLines(input)
        .map((instruction) => instruction.anyContinuationLinesRemoved)
        .join('\n');
}

/**
 * Removes newlines escaped by a backslash.
 * A trailing backslash at the end of a line can be escaped by doubling it.
 *
 * Instruction lines are not trimmed.
 * But instructions that are empty or only contain whitespace are discarded.
 *
 * @param input input string
 * @returns modified input, as a list of strings
 *
 * @see continueLinesFlattened
 */
export function continueLines(input: string): Statement[] {
    const instructions: Statement[] = [];
    let continuePreviousLine = false;

    let currentStatementRaw = '';
    let currentStatementProcessed = '';

    // See https://github.com/obsidian-tasks-group/obsidian-tasks/issues/3137.
    //      Issue #3137 revealed that if the last line of the query ended in
    //      a backslash, this function returned without saved the final
    //      instruction.
    //      The simplest way to prevent this is to add an extra end-of-line
    //      to the end of the query, which will get ignored when not needed:
    const inputWithGuaranteedFinalEOL = input + '\n';
    for (const inputLine of inputWithGuaranteedFinalEOL.split('\n')) {
        const adjustedLine = adjustLine(inputLine, continuePreviousLine);
        if (continuePreviousLine) {
            currentStatementRaw += '\n' + inputLine;
            currentStatementProcessed += ' ' + adjustedLine;
        } else {
            currentStatementRaw = inputLine;
            currentStatementProcessed = adjustedLine;
        }

        // Decide what to do with the next line:
        if (endsWith2Slashes(inputLine)) {
            continuePreviousLine = false;
        } else {
            continuePreviousLine = endsWith1Slash(inputLine);
        }
        if (!continuePreviousLine) {
            if (currentStatementProcessed.trim() !== '') {
                instructions.push(new Statement(currentStatementRaw, currentStatementProcessed));
            }
            currentStatementRaw = '';
            currentStatementProcessed = '';
        }
    }
    return instructions;
}

/**
 * Take an input string and split it into a list of statements.
 *
 * Generally this is similar to splitting the string into lines, but handles line
 * continuations and escape sequences.
 *
 * @param input Input string
 * @returns List of statements
 */
export function splitSourceHonouringLineContinuations(input: string): string[] {
    return continueLines(input).map((instruction: Statement) => instruction.anyContinuationLinesRemoved);
}
