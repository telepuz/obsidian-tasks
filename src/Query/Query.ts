import { getSettings } from '../Config/Settings';
import type { IQuery } from '../IQuery';
import { QueryLayoutOptions, parseQueryShowHideOptions } from '../Layout/QueryLayoutOptions';
import { TaskLayoutOptions, parseTaskShowHideOptions } from '../Layout/TaskLayoutOptions';
import { errorMessageForException } from '../lib/ExceptionTools';
import { logging } from '../lib/logging';
import { expandPlaceholders } from '../Scripting/ExpandPlaceholders';
import { makeQueryContext } from '../Scripting/QueryContext';
import type { Task } from '../Task/Task';
import type { OptionalTasksFile } from '../Scripting/TasksFile';
import { Explainer } from './Explain/Explainer';
import type { Filter } from './Filter/Filter';
import * as FilterParser from './FilterParser';
import type { Grouper } from './Group/Grouper';
import { TaskGroups } from './Group/TaskGroups';
import { QueryResult } from './QueryResult';
import { continueLines } from './Scanner';
import { SearchInfo } from './SearchInfo';
import { Sort } from './Sort/Sort';
import type { Sorter } from './Sort/Sorter';
import { Statement } from './Statement';

let queryInstanceCounter = 0;

export class Query implements IQuery {
    /** Note: source is the raw source, before expanding any placeholders */
    public readonly source: string;

    /** statements contain each source line after processing continuations and placeholders.
     * There may be more statements than lines in the source, if any multi-line query file property values were used. */
    public readonly statements: Statement[] = [];

    public readonly tasksFile: OptionalTasksFile;

    private _limit: number | undefined = undefined;
    private _taskGroupLimit: number | undefined = undefined;

    private readonly _taskLayoutOptions: TaskLayoutOptions = new TaskLayoutOptions();
    private readonly _queryLayoutOptions: QueryLayoutOptions = new QueryLayoutOptions();
    public readonly layoutStatements: Statement[] = [];

    private readonly _filters: Filter[] = [];
    private _error: string | undefined = undefined;
    private readonly _sorting: Sorter[] = [];
    private readonly _grouping: Grouper[] = [];
    private _ignoreGlobalQuery: boolean = false;

    private readonly hideOptionsRegexp = /^(hide|show) +(.*)/i;
    private readonly shortModeRegexp = /^short/i;
    private readonly fullModeRegexp = /^full/i;
    private readonly explainQueryRegexp = /^explain/i;
    private readonly ignoreGlobalQueryRegexp = /^ignore global query/i;

    logger = logging.getLogger('tasks.Query');
    // Used internally to uniquely log each query execution in the console.
    private readonly _queryId: string;

    private readonly limitRegexp = /^limit (groups )?(to )?(\d+)( tasks?)?/i;

    private readonly commentRegexp = /^#.*/;

    constructor(source: string, tasksFile: OptionalTasksFile = undefined) {
        this._queryId = this.generateQueryId(10);

        this.source = source;
        this.tasksFile = tasksFile;

        const anyContinuationLinesRemoved = continueLines(source);

        for (const statement of anyContinuationLinesRemoved) {
            const expandedStatements = this.expandPlaceholders(statement, tasksFile);
            if (this.error !== undefined) {
                // There was an error expanding placeholders.
                return;
            }
            this.statements.push(...expandedStatements);
        }

        for (const statement of this.statements) {
            try {
                this.parseLine(statement);
                if (this.error !== undefined) {
                    return;
                }
            } catch (e) {
                let message;
                if (e instanceof Error) {
                    message = e.message;
                } else {
                    message = 'Unknown error';
                }

                this.setError(message, statement);
                return;
            }
        }
    }

    /**
     * Remove any instructions that are not valid for Global Queries:
     */
    public removeIllegalGlobalQueryInstructions() {
        // It does not make sense to use 'ignore global query'
        // in global queries: the value is ignored, and it would be confusing
        // for 'explain' output to report that it had been supplied:
        this._ignoreGlobalQuery = false;
    }

    public get filePath(): string | undefined {
        return this.tasksFile?.path ?? undefined;
    }

    public get queryId(): string {
        return this._queryId;
    }

    private parseLine(statement: Statement) {
        const line = statement.anyPlaceholdersExpanded;
        switch (true) {
            case this.shortModeRegexp.test(line):
                this._queryLayoutOptions.shortMode = true;
                this.saveLayoutStatement(statement);
                break;
            case this.fullModeRegexp.test(line):
                this._queryLayoutOptions.shortMode = false;
                this.saveLayoutStatement(statement);
                break;
            case this.explainQueryRegexp.test(line):
                this._queryLayoutOptions.explainQuery = true;
                // We intentionally do not explain the 'explain' statement, as it clutters up documentation.
                break;
            case this.ignoreGlobalQueryRegexp.test(line):
                this._ignoreGlobalQuery = true;
                break;
            case this.limitRegexp.test(line):
                this.parseLimit(line);
                break;
            case this.parseSortBy(line, statement):
                break;
            case this.parseGroupBy(line, statement):
                break;
            case this.hideOptionsRegexp.test(line):
                this.parseHideOptions(statement);
                break;
            case this.commentRegexp.test(line):
                // Comment lines are ignored
                break;
            case this.parseFilter(line, statement):
                break;
            default:
                this.setError('do not understand query', statement);
        }
    }

    private formatQueryForLogging() {
        return `
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
${this.source}
<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
`;
    }

    private expandPlaceholders(statement: Statement, tasksFile: OptionalTasksFile): Statement[] {
        const source = statement.anyContinuationLinesRemoved;
        if (source.includes('{{') && source.includes('}}')) {
            if (this.tasksFile === undefined) {
                this._error = `The query looks like it contains a placeholder, with "{{" and "}}"
but no file path has been supplied, so cannot expand placeholder values.
The query is:
${source}`;
                return [statement];
            }
        }

        // TODO Do not complain about any placeholder errors in comment lines
        // TODO Give user error info if they try and put a string in a regex search
        let expandedSource: string = source;
        if (tasksFile) {
            const queryContext = makeQueryContext(tasksFile);
            try {
                expandedSource = expandPlaceholders(source, queryContext);
            } catch (error) {
                if (error instanceof Error) {
                    this._error = error.message;
                } else {
                    this._error = 'Internal error. expandPlaceholders() threw something other than Error.';
                }
                return [statement];
            }
        }

        return this.createStatementsFromExpandedPlaceholders(expandedSource, statement);
    }

    private createStatementsFromExpandedPlaceholders(expandedSource: string, statement: Statement) {
        // Trim and filter empty lines in one step.
        const expandedSourceLines = expandedSource
            .split('\n')
            .map((line) => line.trim())
            .filter((line) => line.length > 0);

        if (expandedSourceLines.length === 1) {
            // Save the single expanded line back into the statement.
            statement.recordExpandedPlaceholders(expandedSourceLines[0]);
            return [statement];
        }

        // Handle multiple-line placeholders.
        return expandedSourceLines.map((expandedSourceLine, index) => {
            const counter = `: statement ${index + 1} after expansion of placeholder`;
            const newStatement = new Statement(
                statement.rawInstruction + counter,
                statement.anyContinuationLinesRemoved + counter,
            );
            newStatement.recordExpandedPlaceholders(expandedSourceLine);
            return newStatement;
        });
    }

    /**
     *
     * Appends {@link q2} to this query.
     *
     * @note At time of writing, this query language appears to play nicely with combining queries.
     *
     * More formally, the concatenation operation on the query language:
     *     * Is closed (concatenating two queries is another valid query)
     *     * Is not commutative (q1.append(q2) !== q2.append(q1))
     *
     * And the semantics of the combination are straight forward:
     *     * Combining two queries appends their filters
     *           (assuming that the filters are pure functions, filter concatenation is commutative)
     *     * Combining two queries appends their sorting instructions. (this is not commutative)
     *     * Combining two queries appends their grouping instructions. (this is not commutative)
     *     * Successive limit instructions overwrite previous ones.
     *
     * @param {Query} q2
     * @return {Query} The combined query
     */
    public append(q2: Query): Query {
        if (this.source === '') return q2;
        if (q2.source === '') return this;
        return new Query(`${this.source}\n${q2.source}`, this.tasksFile);
    }

    /**
     * Generate a text description of the contents of this query.
     *
     * This does not show any global filter and global query.
     * Use {@link explainResults} if you want to see any global query and global filter as well.
     */
    public explainQuery(): string {
        const explainer = new Explainer();
        return explainer.explainQuery(this);
    }

    public get limit(): number | undefined {
        return this._limit;
    }

    public get taskGroupLimit(): number | undefined {
        return this._taskGroupLimit;
    }

    get taskLayoutOptions(): TaskLayoutOptions {
        return this._taskLayoutOptions;
    }

    public get queryLayoutOptions(): QueryLayoutOptions {
        return this._queryLayoutOptions;
    }

    public get filters(): Filter[] {
        return this._filters;
    }

    /**
     * Add a new filter to this Query.
     *
     * At the time of writing, it is intended to allow tests to create filters
     * programatically, for things that can not yet be done via 'filter by function'.
     * @param filter
     */
    public addFilter(filter: Filter) {
        this._filters.push(filter);
    }

    public get sorting() {
        return this._sorting;
    }

    /**
     * Return the {@link Grouper} objects that represent any `group by` instructions in the tasks block.
     */
    public get grouping(): Grouper[] {
        return this._grouping;
    }

    public get error(): string | undefined {
        return this._error;
    }

    private setError(message: string, statement: Statement) {
        this._error = Query.generateErrorMessage(statement, message);
    }

    private static generateErrorMessage(statement: Statement, message: string) {
        if (statement.allLinesIdentical()) {
            return `${message}
Problem line: "${statement.rawInstruction}"`;
        } else {
            return `${message}
Problem statement:
${statement.explainStatement('    ')}
`;
        }
    }

    public get ignoreGlobalQuery(): boolean {
        return this._ignoreGlobalQuery;
    }

    public applyQueryToTasks(tasks: Task[]): QueryResult {
        this.debug(`[search] Executing query: ${this.formatQueryForLogging()}`);

        const searchInfo = new SearchInfo(this.tasksFile, tasks);

        // Custom filter (filter by function) does not report the instruction line in any exceptions,
        // for performance reasons. So we keep track of it here.
        let possiblyBrokenStatement: Statement | undefined = undefined;
        try {
            this.filters.forEach((filter) => {
                possiblyBrokenStatement = filter.statement;
                tasks = tasks.filter((task) => filter.filterFunction(task, searchInfo));
            });
            possiblyBrokenStatement = undefined;

            const { debugSettings } = getSettings();
            const tasksSorted = debugSettings.ignoreSortInstructions ? tasks : Sort.by(this.sorting, tasks, searchInfo);
            const tasksSortedLimited = tasksSorted.slice(0, this.limit);

            const taskGroups = new TaskGroups(this.grouping, tasksSortedLimited, searchInfo);

            if (this._taskGroupLimit !== undefined) {
                taskGroups.applyTaskLimit(this._taskGroupLimit);
            }

            return new QueryResult(taskGroups, tasksSorted.length);
        } catch (e) {
            const description = 'Search failed';
            let message = errorMessageForException(description, e);

            if (possiblyBrokenStatement) {
                message = Query.generateErrorMessage(possiblyBrokenStatement, message);
            }
            return QueryResult.fromError(message);
        }
    }

    private parseHideOptions(statement: Statement): void {
        const line = statement.anyPlaceholdersExpanded;
        const hideOptionsMatch = line.match(this.hideOptionsRegexp);
        if (hideOptionsMatch === null) {
            return;
        }
        const hide = hideOptionsMatch[1].toLowerCase() === 'hide';
        const option = hideOptionsMatch[2].toLowerCase();

        if (parseQueryShowHideOptions(this._queryLayoutOptions, option, hide)) {
            this.saveLayoutStatement(statement);
            return;
        }
        if (parseTaskShowHideOptions(this._taskLayoutOptions, option, !hide)) {
            this.saveLayoutStatement(statement);
            return;
        }
        this.setError('do not understand hide/show option', new Statement(line, line));
    }

    private saveLayoutStatement(statement: Statement) {
        this.layoutStatements.push(statement);
    }

    private parseFilter(line: string, statement: Statement) {
        const filterOrError = FilterParser.parseFilter(line);
        if (filterOrError != null) {
            if (filterOrError.filter) {
                // Overwrite the filter's statement, to preserve details of any
                // continuation lines and placeholder expansions.
                filterOrError.filter.setStatement(statement);

                this._filters.push(filterOrError.filter);
            } else {
                this.setError(filterOrError.error ?? 'Unknown error', statement);
            }
            return true;
        }
        return false;
    }

    private parseLimit(line: string): void {
        const limitMatch = line.match(this.limitRegexp);
        if (limitMatch === null) {
            this.setError('do not understand query limit', new Statement(line, line));
            return;
        }

        // limitMatch[3] is per regex always digits and therefore parsable.
        const limitFromLine = Number.parseInt(limitMatch[3], 10);

        if (limitMatch[1] !== undefined) {
            this._taskGroupLimit = limitFromLine;
        } else {
            this._limit = limitFromLine;
        }
    }

    private parseSortBy(line: string, statement: Statement): boolean {
        const sortingMaybe = FilterParser.parseSorter(line);
        if (sortingMaybe) {
            sortingMaybe.setStatement(statement);
            this._sorting.push(sortingMaybe);
            return true;
        }
        return false;
    }

    /**
     * Parsing of `group by` lines, for grouping that is implemented in the {@link Field}
     * classes.
     *
     * @param line
     * @param statement
     * @private
     */
    private parseGroupBy(line: string, statement: Statement): boolean {
        const groupingMaybe = FilterParser.parseGrouper(line);
        if (groupingMaybe) {
            groupingMaybe.setStatement(statement);
            this._grouping.push(groupingMaybe);
            return true;
        }
        return false;
    }

    /**
     * Creates a unique ID for correlation of console logging.
     *
     * @private
     * @param {number} length
     * @return {*}  {string}
     */
    private generateQueryId(length: number): string {
        queryInstanceCounter += 1;
        return queryInstanceCounter.toString().padStart(length, '0');
    }

    public debug(message: string, objects?: any): void {
        this.logger.debugWithId(this._queryId, `"${this.filePath}": ${message}`, objects);
    }
}
