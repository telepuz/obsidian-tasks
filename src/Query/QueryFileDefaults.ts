import type { OptionalTasksFile, TasksFile } from '../Scripting/TasksFile';
import { Query } from './Query';

// Enum for handler types
enum Handler {
    Instruction = 'instruction',
    ShowAndHide = 'showAndHide',
    AddValue = 'addValue',
}

// Note: This file is excluded from SonarCloud duplication-checks,
//       as the duplication here provides clarity.
// Instructions are listed in the order that items are displayed in Tasks search results
const queryProperties = [
    {
        name: 'TQ_explain',
        type: 'checkbox',
        display: 'explain',
        handler: Handler.Instruction,
        trueValue: 'explain',
        falseValue: '',
    },
    {
        name: 'TQ_short_mode',
        type: 'checkbox',
        display: 'short mode',
        handler: Handler.Instruction,
        trueValue: 'short mode',
        falseValue: 'full mode',
    },
    {
        name: 'TQ_show_tree',
        type: 'checkbox',
        display: 'tree',
        handler: Handler.ShowAndHide,
    },

    // Fields that appear before date values:
    {
        name: 'TQ_show_tags',
        type: 'checkbox',
        display: 'tags',
        handler: Handler.ShowAndHide,
    },
    {
        name: 'TQ_show_id',
        type: 'checkbox',
        display: 'id',
        handler: Handler.ShowAndHide,
    },
    {
        name: 'TQ_show_depends_on',
        type: 'checkbox',
        display: 'depends on',
        handler: Handler.ShowAndHide,
    },
    {
        name: 'TQ_show_priority',
        type: 'checkbox',
        display: 'priority',
        handler: Handler.ShowAndHide,
    },
    {
        name: 'TQ_show_recurrence_rule',
        type: 'checkbox',
        display: 'recurrence rule',
        handler: Handler.ShowAndHide,
    },
    {
        name: 'TQ_show_on_completion',
        type: 'checkbox',
        display: 'on completion',
        handler: Handler.ShowAndHide,
    },

    // Date fields:
    {
        name: 'TQ_show_created_date',
        type: 'checkbox',
        display: 'created date',
        handler: Handler.ShowAndHide,
    },
    {
        name: 'TQ_show_start_date',
        type: 'checkbox',
        display: 'start date',
        handler: Handler.ShowAndHide,
    },
    {
        name: 'TQ_show_scheduled_date',
        type: 'checkbox',
        display: 'scheduled date',
        handler: Handler.ShowAndHide,
    },
    {
        name: 'TQ_show_due_date',
        type: 'checkbox',
        display: 'due date',
        handler: Handler.ShowAndHide,
    },
    {
        name: 'TQ_show_cancelled_date',
        type: 'checkbox',
        display: 'cancelled date',
        handler: Handler.ShowAndHide,
    },
    {
        name: 'TQ_show_done_date',
        type: 'checkbox',
        display: 'done date',
        handler: Handler.ShowAndHide,
    },

    // Elements of query results:
    {
        name: 'TQ_show_urgency',
        type: 'checkbox',
        display: 'urgency',
        handler: Handler.ShowAndHide,
    },
    {
        name: 'TQ_show_backlink',
        type: 'checkbox',
        display: 'backlink',
        handler: Handler.ShowAndHide,
    },
    {
        name: 'TQ_show_edit_button',
        type: 'checkbox',
        display: 'edit button',
        handler: Handler.ShowAndHide,
    },
    {
        name: 'TQ_show_postpone_button',
        type: 'checkbox',
        display: 'postpone button',
        handler: Handler.ShowAndHide,
    },
    {
        name: 'TQ_show_task_count',
        type: 'checkbox',
        display: 'task count',
        handler: Handler.ShowAndHide,
    },
    {
        name: 'TQ_extra_instructions',
        type: 'text',
        handler: Handler.AddValue,
    },
];

/**
 * Construct query instructions from Obsidian properties in the query file
 */
export class QueryFileDefaults {
    public source(queryFile: OptionalTasksFile) {
        if (!queryFile) {
            return '';
        }

        const instructions = queryProperties.map((prop) => this.generateInstruction(queryFile, prop));
        return instructions.filter((i) => i !== '').join('\n');
    }

    private generateInstruction(queryFile: TasksFile, prop: any) {
        const hasProperty = queryFile.hasProperty(prop.name);
        const value = queryFile.property(prop.name);
        switch (prop.handler) {
            case Handler.Instruction:
                return (hasProperty && (value ? prop.trueValue : prop.falseValue)) || '';
            case Handler.ShowAndHide:
                return (hasProperty && (value ? 'show ' + prop.display : 'hide ' + prop.display)) || '';
            case Handler.AddValue:
                return hasProperty ? value || '' : '';
            default:
                throw new Error('Unknown handler type: ' + prop.handler + '.');
        }
    }

    public query(queryFile: OptionalTasksFile) {
        return new Query(this.source(queryFile), queryFile);
    }

    public allPropertyNames(): string[] {
        return queryProperties.map((prop) => prop.name);
    }

    public allPropertyNamesSorted(): string[] {
        return this.allPropertyNames().sort((a, b) => a.localeCompare(b));
    }

    /**
     * Return the type of the property with the specified name.
     *
     * @param {string} propertyName - The name of the property to retrieve the type for.
     * @return {string | undefined} The type of the property if found, or undefined if the property is not found.
     */
    public propertyType(propertyName: string): string | undefined {
        return queryProperties.find((prop) => prop.name === propertyName)?.type ?? undefined;
    }

    /**
     * Return text that creates MetaBind widgets for users to edit query file defaults.
     */
    public metaBindPluginWidgets() {
        // This is initially hard-coded, though I intend to machine-generate it eventually.
        // Its text is embedded in the test vault and in the user guide.
        return `
short mode: \`INPUT[toggle:TQ_short_mode]\`
tree: \`INPUT[toggle:TQ_show_tree]\`
tags: \`INPUT[toggle:TQ_show_tags]\`
id: \`INPUT[toggle:TQ_show_id]\` depends on: \`INPUT[toggle:TQ_show_depends_on]\`
priority: \`INPUT[toggle:TQ_show_priority]\`
recurrence rule: \`INPUT[toggle:TQ_show_recurrence_rule]\` on completion: \`INPUT[toggle:TQ_show_on_completion]\`
start date: \`INPUT[toggle:TQ_show_start_date]\` scheduled date: \`INPUT[toggle:TQ_show_scheduled_date]\` due date: \`INPUT[toggle:TQ_show_due_date]\`
created date: \`INPUT[toggle:TQ_show_created_date]\` cancelled date: \`INPUT[toggle:TQ_show_cancelled_date]\` done date: \`INPUT[toggle:TQ_show_done_date]\`
urgency: \`INPUT[toggle:TQ_show_urgency]\`
backlink: \`INPUT[toggle:TQ_show_backlink]\`
edit button: \`INPUT[toggle:TQ_show_edit_button]\` postpone button: \`INPUT[toggle:TQ_show_postpone_button]\`
task count: \`INPUT[toggle:TQ_show_task_count]\`
extra instructions: \`INPUT[textArea:TQ_extra_instructions]\`
explain: \`INPUT[toggle:TQ_explain]\`
`;
    }
}
