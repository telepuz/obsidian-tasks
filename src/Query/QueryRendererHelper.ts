import type { GlobalFilter } from '../Config/GlobalFilter';
import type { GlobalQuery } from '../Config/GlobalQuery';
import type { OptionalTasksFile } from '../Scripting/TasksFile';
import { Query } from './Query';
import { Explainer } from './Explain/Explainer';
import { QueryFileDefaults } from './QueryFileDefaults';

/**
 * @summary
 * This file contains utilities used by {@link QueryRenderer} that should actually
 * be in that file. But the file {@link QueryRenderer} is in depends on dependencies
 * that aren't available during automated testing, and would make it impossible to run
 * those tests on the code that is instead housed here.
 */

/**
 * Explains a query rendered by {@link QueryRenderer}
 *
 * Specifically, returns a formatted string:
 *     * Explains whether a global filter is in use
 *     * Explains whether the global query if it's in use
 *     * Explains the query described by {@link source}
 *
 * @param {string} source The source of the task block to explain
 * @param {GlobalFilter} globalFilter The global filter. In `src/`, generally pass in {@link GlobalFilter.getInstance}
 * @param {GlobalQuery} globalQuery The global query. In `src/`, generally pass in {@link GlobalQuery.getInstance}
 * @param {OptionalTasksFile} tasksFile The location of the task block, if known
 * @returns {string}
 */
export function explainResults(
    source: string,
    globalFilter: GlobalFilter,
    globalQuery: GlobalQuery,
    tasksFile: OptionalTasksFile = undefined,
): string {
    let result = '';

    if (!globalFilter.isEmpty()) {
        result += `Only tasks containing the global filter '${globalFilter.get()}'.\n\n`;
    }

    const explainer = new Explainer('  ');
    function explainQuery(label: string, query: Query) {
        return `${label}:\n\n${explainer.explainQuery(query)}`;
    }

    const tasksBlockQuery = new Query(source, tasksFile);
    const queryFileDefaultsQuery = new QueryFileDefaults().query(tasksFile);

    const ignoreGlobalQuery = tasksBlockQuery.ignoreGlobalQuery || queryFileDefaultsQuery.ignoreGlobalQuery;
    if (!ignoreGlobalQuery) {
        if (globalQuery.hasInstructions()) {
            const globalQueryQuery = globalQuery.query(tasksFile);
            result += explainQuery('Explanation of the global query', globalQueryQuery) + '\n';
        }
    }

    if (queryFileDefaultsQuery.source !== '') {
        const intro = "Explanation of the Query File Defaults (from properties/frontmatter in the query's file)";
        result += explainQuery(intro, queryFileDefaultsQuery) + '\n';
    }

    result += explainQuery('Explanation of this Tasks code block query', tasksBlockQuery);

    return result;
}

/**
 * Creates the actual query that {@link QueryRenderChild} will actually execute against the task list.
 *
 * This query is the result of joining the global query with the query in the task block
 *
 * @param {string} source The query source from the task block
 * @param globalQuery
 * @param {OptionalTasksFile} tasksFile The path to the file containing the query, if available.
 * @returns {Query} The query to execute
 */
export function getQueryForQueryRenderer(
    source: string,
    globalQuery: GlobalQuery,
    tasksFile: OptionalTasksFile,
): Query {
    // Construct query from any recognised properties:
    const queryFileDefaultsQuery: Query = new QueryFileDefaults().query(tasksFile);

    // Prepend the QueryFileDefaults query - from properties - to the tasks block source:
    const tasksBlockQuery = queryFileDefaultsQuery.append(new Query(source, tasksFile));

    if (tasksBlockQuery.ignoreGlobalQuery) {
        return tasksBlockQuery;
    }

    return globalQuery.query(tasksFile).append(tasksBlockQuery);
}
