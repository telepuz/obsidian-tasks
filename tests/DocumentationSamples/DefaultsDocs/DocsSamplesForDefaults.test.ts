import type { Pos } from 'obsidian';

import { verify, verifyAsJson } from 'approvals/lib/Providers/Jest/JestApprovals';
import { getTasksFileFromMockData } from '../../TestingTools/MockDataHelpers';
import query_file_defaults_all_options_null from '../../Obsidian/__test_data__/query_file_defaults_all_options_null.json';
import query_file_defaults_all_options_true from '../../Obsidian/__test_data__/query_file_defaults_all_options_true.json';
import query_file_defaults_short_mode from '../../Obsidian/__test_data__/query_file_defaults_short_mode.json';
import { verifyWithFileExtension } from '../../TestingTools/ApprovalTestHelpers';
import { verifyMarkdown, verifyMarkdownForDocs } from '../../TestingTools/VerifyMarkdown';
import { QueryFileDefaults } from '../../../src/Query/QueryFileDefaults';

function extractFrontmatter(data: any) {
    const queryFile = getTasksFileFromMockData(data);
    const pos: Pos | undefined = queryFile.cachedMetadata.frontmatterPosition;
    return data.fileContents.slice(pos?.start.offset ?? 0, pos?.end.offset ?? 0);
}

describe('DocsSamplesForDefaults', () => {
    it('supported-properties-empty', () => {
        const frontmatter = extractFrontmatter(query_file_defaults_all_options_null);

        // Make sure that any trailing spaces have been removed from the
        // properties in query_file_defaults_all_options_null.md, to avoid
        // fighting with spaces at end of line when the approved file from this
        // test is embedded in the user docs.
        expect(frontmatter).not.toContain(': ');

        verifyWithFileExtension(frontmatter, '.yaml');
    });

    it('supported-properties-full', () => {
        verifyWithFileExtension(extractFrontmatter(query_file_defaults_all_options_true), '.yaml');
    });

    describe('demo-short-mode', () => {
        // Load query_file_defaults_short_mode.json
        const data = query_file_defaults_short_mode;

        it('yaml', () => {
            // Extract the frontmatter to a file, for docs
            verifyWithFileExtension(extractFrontmatter(data), '.yaml');
        });

        it('instructions', () => {
            // Create the instruction from it, for docs
            const tasksFile = getTasksFileFromMockData(data);
            const generatedSource = new QueryFileDefaults().source(tasksFile);
            verify(generatedSource);
        });
    });

    it('meta-bind-widgets-include', () => {
        verifyMarkdownForDocs(new QueryFileDefaults().metaBindPluginWidgets());
    });

    it('meta-bind-widgets-snippet', () => {
        verifyMarkdown(new QueryFileDefaults().metaBindPluginWidgets());
    });

    it('fake-types.json', () => {
        const queryFileDefaults = new QueryFileDefaults();
        const allPropertyNamesSorted = queryFileDefaults.allPropertyNamesSorted();

        // Generate an object representing QueryFileDefaults in Obsidian's types.json.
        const types: { [key: string]: string } = allPropertyNamesSorted.reduce(
            (acc: { [key: string]: string }, propName) => {
                const propType = queryFileDefaults.propertyType(propName);
                if (propType !== undefined) {
                    acc[propName] = propType;
                }
                return acc;
            },
            {},
        );
        const generatedObject = { types };

        verifyAsJson(generatedObject);
    });
});
