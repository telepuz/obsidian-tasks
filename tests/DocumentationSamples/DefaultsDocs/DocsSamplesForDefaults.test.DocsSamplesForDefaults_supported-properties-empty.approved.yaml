---
TQ_explain:
TQ_extra_instructions:
TQ_short_mode:
TQ_show_backlink:
TQ_show_cancelled_date:
TQ_show_created_date:
TQ_show_depends_on:
TQ_show_done_date:
TQ_show_due_date:
TQ_show_edit_button:
TQ_show_id:
TQ_show_on_completion:
TQ_show_postpone_button:
TQ_show_priority:
TQ_show_recurrence_rule:
TQ_show_scheduled_date:
TQ_show_start_date:
TQ_show_tags:
TQ_show_task_count:
TQ_show_tree:
TQ_show_urgency:
---
